pytoda/datasets/polymer_dataset.py

```diff
@@ -61,113 +64,153 @@ def __init__(
+                all_hs_explicit=self.all_hs_explicits[0],
+                randomize=self.randomizes[0],
+                remove_bonddir=self.remove_bonddirs[0],
+                remove_chirality=self.remove_chiralitys[0],
+                selfies=self.selfies[0],
+                sanitize=self.sanitize[0],
+                device=device,
+                add_start_and_stop=True,
+            )
+            for index, entity in enumerate(entity_names):
+                self.smiles_language.set_smiles_transforms(
+                    entity,
+                    canonical=self.canonicals[index],
+                    augment=self.augments[index],
+                    kekulize=self.kekulizes[index],
+                    all_bonds_explicit=self.all_bonds_explicits[index],
+                    all_hs_explicit=self.all_hs_explicits[index],
+                    remove_bonddir=self.remove_bonddirs[index],
+                    remove_chirality=self.remove_chiralitys[index],
+                    selfies=self.selfies[index],
+                    sanitize=self.sanitize[index],
+                )
+                # set_encoding_transforms only after adding smiles
+
+        else:
+            self.smiles_language = smiles_language
+
+        self.entities = self.smiles_language.entities
+        self.datasets = [
             SMILESDataset(
-                smi_filepaths[index],
+                smi_filepath,
                 name=self.entities[index],
-                smiles_language=self.smiles_language,
-                padding=self.paddings[index],
-                padding_length=self.padding_lengths[index],
-                canonical=self.canonicals[index],
-                augment=self.augments[index],
-                kekulize=self.kekulizes[index],
-                all_bonds_explicit=self.all_bonds_explicits[index],
-                all_hs_explicit=self.all_hs_explicits[index],
-                randomize=self.randomizes[index],
-                remove_bonddir=self.remove_bonddirs[index],
-                remove_chirality=self.remove_chiralitys[index],
-                selfies=self.selfies[index],
-                sanitize=self.sanitize[index],
-                device=device
-            ) for index in range(len(smi_filepaths))
+            ) for index, smi_filepath in enumerate(smi_filepaths)
         ]
-        """
-        Push the Polymer language configuration down to the smiles language
-        object associated to the dataset and to the tokenizer that use this
-        object.
-        """
-        for dataset in self._datasets:
 
-            dataset.smiles_language = deepcopy(self.smiles_language)
-            dataset.smiles_language.update_entity(dataset.name)
-            tokenizer_index = [
-                i for i, t in enumerate(dataset._dataset.transform.transforms)
-                if isinstance(t, SMILESToTokenIndexes)
-            ]
-            if len(tokenizer_index) > 0:
```

Comment:
can you briefly explain why this hack is not needed anymore @C-nit ?

Reply:
2 things going on:  

Each dataset is only returning SMILES now, and is language independent.
where before the datasets each had a language that was shared, there is a single language outside that is used now (building the vocab over all datasets)

with the split in smiles vs encoding transforms, SMILESToTokenIndexes is not used anymore, but the languages ...to_token_indexes methods directly. So no need to hack into the tranforms list and change the instance.